app/models/service_template_transformation_plan_task.rb

```diff
@@ -256,12 +256,12 @@ def calculate_network_mappings
       source_network = nic.lan
       destination_network = transformation_destination(source_network)
       raise "[#{source.name}] NIC #{nic.device_name} [#{source_network.name}] has no mapping." if destination_network.nil?
-      raise "[#{source.name}] NIC #{nic.device_name} [#{source_network.name}] has an empty IP address." if nic.network.try(:ipaddress).nil?
+      raise "[#{source.name}] NIC #{nic.device_name} [#{source_network.name}] has an empty IP address." if nic.network.try(:ipaddress).nil? && destination_ems.emstype == 'openstack'
```

Comment:
:-1: really don't like hard coding this into the common base class

Reply:
The validation is specific to IMS. That's an IMS limitation due to the way CloudForms stores the IP addresses for a single NIC. When a NIC has a number of IPv4 or IPv6 addresses greater than 1, the Network object associated to the Nic object can have a nil IP address.

BTW, we also handle the various providers using `send` with method names including the provider name. How different is that ? Shoud I split `network_mappings` into `network_mappings_rhevm` and `network_mappings_openstack` ?

Reply:
@fdupont-redhat Let's create some scaffolding in the provider repo and put it there. I already have `ServiceTemplateTransformationPlan` at https://github.com/ManageIQ/manageiq-providers-openstack/blob/master/app/models/manageiq/providers/openstack/cloud_manager/service_template_transformation_plan.rb.

So, how about we create ServiceTemplateTransformationPlanTask there as well, and then put our rules in there. Let's establish a common interface method. Then have the core method use that.

Reply:
@bzwei and I looked around yesterday. I *think* this will "just work" if you subclass it and redefine it in the subclass, as rails should automagically set the type properly. @agrare am I wrong?

Reply:
@djberg96 exactly as long as you actually have an instance of `ManageIQ::Providers::Openstack::CloudManager::ServiceTemplateTransformationPlan` it'll call into that method first.

Typically we'd define
```
def calculate_network_mappings
  super # call the base code
  additional_super_awesome_validations_only_for_openstack
end
```

Reply:
There's some magic happening at https://github.com/ManageIQ/manageiq/blob/master/app/models/miq_request.rb#L341 inside MiqRequest, but I *think* it will still work.

I would normally call `super`, but since it's a loop, he might be better off just redefining the method completely, otherwise he'll have to loop again. That, or create some sort of common interface method for a validating check, but that's more work.

Reply:
Reading virt-v2v-wrapper code, it appears that a missing IP address is not a problem. What fails is a IP address with `nil` value, as virt-v2v-wrapper will add the `None` literal to the `openstack` command. I discussed with Marco and he agreed that there are valid use cases for missing IP address, regardless of [BZ#1657126](https://bugzilla.redhat.com/show_bug.cgi?id=1657126), so we should not forcefully fail but rather prevent virt-v2v-wrapper failure. I updated the PR to reflect that and the `network_mappings` is again generic.

Reply:
I've updated the PR title and description.